File: techpath-backend/app/schemas/student_portal.py
"""Schemas for the post-session student materials portal.

Separate from ``schemas/classroom.py`` on purpose: that file is the live-session,
no-account student experience (join code + short-lived token); this one is the
durable, Firebase-authenticated "come back later" experience. Different identity,
different lifetime, different access rule — keeping them apart keeps either from
growing an accidental dependency on the other's assumptions.
"""

from datetime import datetime
from typing import Any, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.classroom import RecordingView
from app.schemas.training import LectureAssetResponse


class StudentLoginResponse(BaseModel):
    display_name: str
    email: Optional[str] = None


class StudentSessionSummary(BaseModel):
    session_id: int
    title: Optional[str] = None
    batch_name: str
    module_title: Optional[str] = None
    session_date: Optional[datetime] = None
    published_at: datetime

    model_config = ConfigDict(from_attributes=False)


class StudentSessionListResponse(BaseModel):
    sessions: List[StudentSessionSummary]


class StudentSessionMaterialsResponse(BaseModel):
    session_id: int
    title: Optional[str] = None
    batch_name: str
    module_title: Optional[str] = None
    published_at: datetime
    assets: List[LectureAssetResponse]
    # Null when the session never had live media (chat/poll-only class) — distinct from
    # RecordingView.status == 'processing', which means media happened and the replay
    # just isn't ready yet.
    recording: Optional[RecordingView] = None


# ---------------------------------------------------------------------------
# Graded quizzes
# ---------------------------------------------------------------------------


class QuizAttemptSubmission(BaseModel):
    """A student's answers for one quiz.

    Answers only — deliberately no score field. The server grades against the stored
    answer key, and a client-supplied score would be worthless even if it were sent.
    Element ``i`` is the selected option index for question ``i``; per-element range
    checking happens in the grading service, which is the only thing that knows how
    many options each question has.
    """

    answers: List[Any] = Field(
        ...,
        description="Selected option index per question, positionally aligned to the quiz.",
    )


class QuizQuestionFeedback(BaseModel):
    """Per-question result. The only path by which a student receives an answer key,
    and only for an attempt they have already submitted."""

    index: int
    your_answer: int
    correct_index: Optional[int] = None
    is_correct: bool
    explanation: Optional[str] = None


class QuizAttemptResult(BaseModel):
    attempt_id: int
    attempt_number: int
    score: int
    total_questions: int
    percentage: float
    passed: bool
    pass_mark: float
    attempted_at: datetime
    # Lets the portal reveal the next material item straight away rather than
    # refetching progress after a passing submission.
    unlocked_next: bool
    questions: List[QuizQuestionFeedback]


class StudentProgressItem(BaseModel):
    """One material item's state for one student.

    ``passed`` is null for non-quiz items — they have nothing to pass, which is
    different from a quiz that has been failed.
    """

    asset_id: int
    index: int
    is_quiz: bool
    passed: Optional[bool] = None
    locked: bool
    best_score: Optional[int] = None
    total_questions: Optional[int] = None
    attempt_count: Optional[int] = None


class StudentProgressResponse(BaseModel):
    session_id: int
    # Index of the first quiz without a passing attempt; equals len(items) when there
    # is none. The quiz *at* this index is reachable — everything after it is not.
    first_locked_index: int
    items: List[StudentProgressItem]

File: techpath-frontend/src/types/studentPortal.ts
import type { ClassroomAsset } from './classroom';

/**
 * The student materials portal — a durable, Google-authenticated identity system,
 * deliberately separate from the live classroom's 6-digit-code + short-lived-token
 * flow (see classroom.ts). A student who attended a published session signs in here
 * with a real Firebase/Google account, days or weeks later, to view/download whatever
 * the trainer has published for that session.
 *
 * The one thing genuinely shared with the live classroom is the asset shape itself —
 * both surfaces serialize assets via the same backend `asset_to_response()`, so
 * `ClassroomAsset` is reused as-is here rather than duplicated.
 */

export interface StudentLoginResponse {
  display_name: string;
  email: string | null;
}

export interface StudentSessionSummary {
  session_id: number;
  title: string | null;
  batch_name: string;
  module_title: string | null;
  session_date: string;
  published_at: string;
}

export interface StudentSessionListResponse {
  sessions: StudentSessionSummary[];
}

/** The class's recorded replay, if it ever had live media. `watch_url` is always
 *  present once this object exists, but only actually playable once `status ===
 *  'ready'` — show a "still processing" state until then. */
export interface RecordingView {
  status: 'processing' | 'ready' | 'failed';
  watch_url: string | null;
}

export interface StudentSessionMaterialsResponse {
  session_id: number;
  title: string | null;
  batch_name: string;
  module_title: string | null;
  published_at: string;
  assets: ClassroomAsset[];
  recording: RecordingView | null;
}

/* -------------------------------------------------------------------------
 * Graded quizzes
 *
 * QuizAttemptResult / QuizQuestionFeedback are declared in classroom.ts and
 * re-exported here: both the portal and the live classroom submit quizzes, so
 * the shape belongs on the shared base module rather than on this one.
 * ---------------------------------------------------------------------- */

export type { QuizAttemptResult, QuizQuestionFeedback } from './classroom';

export interface StudentProgressItem {
  asset_id: number;
  index: number;
  is_quiz: boolean;
  /** Null for non-quiz items — they have nothing to pass, which is different
   *  from a quiz that has been failed. */
  passed: boolean | null;
  locked: boolean;
  best_score: number | null;
  total_questions: number | null;
  attempt_count: number | null;
}

export interface StudentProgressResponse {
  session_id: number | null;
  module_id: number | null;
  first_locked_index: number;
  items: StudentProgressItem[];
}


// ---------------------------------------------------------------------------
// Self-paced courses
// ---------------------------------------------------------------------------

export interface SelfPacedModuleSummary {
  module_id: number;
  title: string;
  description: string | null;
  display_order: number;
  estimated_minutes: number | null;
  asset_count: number;
  quiz_count: number;
  started: boolean;
  completed: boolean;
  last_asset_index: number;
}

export interface SelfPacedCourseSummary {
  program_id: number;
  title: string;
  slug: string;
  summary: string | null;
  cover_image: string | null;
  delivery_mode: string;
  level: string | null;
  duration: string | null;
  batch_name: string;
  module_count: number;
  completed_modules: number;
  total_assets: number;
}

export interface SelfPacedCourseListResponse {
  courses: SelfPacedCourseSummary[];
}

export interface SelfPacedCourseDetailResponse {
  program_id: number;
  title: string;
  slug: string;
  summary: string | null;
  description: string | null;
  cover_image: string | null;
  delivery_mode: string;
  level: string | null;
  duration: string | null;
  batch_name: string;
  modules: SelfPacedModuleSummary[];
}

export interface SelfPacedModuleMaterialsResponse {
  program_id: number;
  module_id: number;
  module_title: string;
  program_title: string;
  batch_name: string;
  assets: ClassroomAsset[];
}
